internal/rbd/replicationcontrollerserver.go

```diff
@@ -0,0 +1,351 @@
+	defer cs.cleanup(rbdVol, cr)
+	if err != nil {
+		return nil, err
+	}
+
+	// extract the force option
+	force, err := getForceOption(ctx, req.GetParameters())
+	if err != nil {
+		return nil, err
+	}
+
+	mirroringInfo, err := rbdVol.getImageMirroringInfo(ctx)
+	if err != nil {
+		err = fmt.Errorf("failed to get mirroring info of %s with error %s", rbdVol.String(), err)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	if mirroringInfo.State != librbd.MirrorImageEnabled {
+		return nil, status.Errorf(codes.InvalidArgument, "image is in %d Mode", mirroringInfo.State)
+	}
+
+	// promote secondary to primary
+	if !mirroringInfo.Primary {
+		err = rbdVol.promoteImage(ctx, force)
+		if err != nil {
+			util.ErrorLog(ctx, "failed to promote mirroring of %s with error %s", err.Error())
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+	}
+
+	return &replication.PromoteVolumeResponse{}, nil
+}
+
+// DemoteVolume demotes the volume as secondary.
+func (cs *ControllerServer) DemoteVolume(ctx context.Context,
+	req *replication.DemoteVolumeRequest,
+) (*replication.DemoteVolumeResponse, error) {
+	// get rbdVolume from VolumeID
+	rbdVol, cr, err := cs.getVolumeFromID(ctx, req.GetVolumeId(), req.GetSecrets())
+	defer cs.cleanup(rbdVol, cr)
+	if err != nil {
+		return nil, err
+	}
+
+	mirroringInfo, err := rbdVol.getImageMirroringInfo(ctx)
+	if err != nil {
+		err = fmt.Errorf("failed to get mirroring info of %s with error %s", rbdVol.String(), err)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	if mirroringInfo.State != librbd.MirrorImageEnabled {
+		return nil, status.Errorf(codes.InvalidArgument, "image is in %d Mode", mirroringInfo.State)
+	}
+
+	// demote image to secondary
+	if mirroringInfo.Primary {
+		err = rbdVol.demoteImage(ctx)
+		if err != nil {
+			util.ErrorLog(ctx, "failed to demote mirroring of %s with error %s", err.Error())
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+	}
+	return &replication.DemoteVolumeResponse{}, nil
+}
+
+// ResyncVolume resync the volume.
+func (cs *ControllerServer) ResyncVolume(ctx context.Context,
+	req *replication.ResyncVolumeRequest,
+) (*replication.ResyncVolumeResponse, error) {
+	// get rbdVolume from VolumeID
+	rbdVol, cr, err := cs.getVolumeFromID(ctx, req.GetVolumeId(), req.GetSecrets())
+	defer cs.cleanup(rbdVol, cr)
+	if err != nil {
+		return nil, err
+	}
+
+	mirroringInfo, err := rbdVol.getImageMirroringInfo(ctx)
+	if err != nil {
+		// TODO check if the image is not found return valid error message
+		// in case of Resync the image will get deleted and gets recreated and
+		// it takes time for this operation
+		err = fmt.Errorf("failed to get mirroring info of %s with error %s", rbdVol.String(), err)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	if mirroringInfo.State != librbd.MirrorImageEnabled {
+		return nil, status.Error(codes.InvalidArgument, "image mirroring is not enabled")
```

Comment:
not enabled where? in the Ceph Pool, or on the image? detailed errors please

Reply:
as we are doing it image level.  we are returning `image mirroring is not enabled` any suggestion on the error message?

Reply:
this is fine, sorry for the confusion